File: Day Challenge/20230713 Course Schedule/207. Course Schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for [a, b] in prerequisites:
            graph[b].append(a)
            indegree[a] += 1

        q = deque([])

        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)

        nodesVisited = 0

        while q:
            node = q.popleft()
            nodesVisited += 1
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                if (indegree[neighbor] == 0):
                    q.append(neighbor)

        return nodesVisited == numCourses

File: Day Challenge/20230713 Course Schedule/207. Course Schedule.js
var canFinish = function (numCourses, prerequisites) {
  let adj = {};
  let indegree = Array(numCourses).fill(0);

  for (let i = 0; i < numCourses; i++) {
    adj[i] = [];
  }

  for (const [a, b] of prerequisites) {
    adj[b].push(a);
    indegree[a]++;
  }

  const queue = [];
  for (let i = 0; i < numCourses; i++) {
    if (indegree[i] === 0) queue.push(i);
  }

  let nodesVisited = 0;
  while (queue.length) {
    node = queue.shift();
    nodesVisited++;

    for (let neighbor of adj[node]) {
      indegree[neighbor]--;
      if (indegree[neighbor] === 0) {
        queue.push(neighbor);
      }
    }
  }

  return nodesVisited === numCourses;
};
